include/deal.II/matrix_free/fe_remote_evaluation.h

```diff
@@ -0,0 +1,768 @@
+    for (const auto &cell : cell_iterator_range)
+      {
+        for (const auto &f : cell->face_indices())
+          {
+            const unsigned int face_index =
+              cell_ptrs[cell->active_cell_index()] + f;
+
+            face_ptrs[face_index + 1] =
+              face_ptrs[face_index] +
+              quadrature_vector[cell->active_cell_index()][f].size();
+          }
+      }
+  }
+
+  /**
+   * Fill the fields stored in PrecomputedFEEvaluationData.
+   */
+  template <int n_components,
+            typename PrecomputedFEEvaluationDataType,
+            typename MeshType,
+            typename VectorType>
+  void
+  update_ghost_values(
+    PrecomputedFEEvaluationDataType       &dst,
+    const MeshType                        &mesh,
+    const VectorType                      &src,
+    const EvaluationFlags::EvaluationFlags eval_flags,
+    const unsigned int                     first_selected_component,
+    const VectorTools::EvaluationFlags::EvaluationFlags vec_flags) const
+  {
+    const bool has_ghost_elements = src.has_ghost_elements();
+
+    if (has_ghost_elements == false)
+      src.update_ghost_values();
+
+
+    for (const auto &communication_object : communication_objects)
+      {
+        if (eval_flags & EvaluationFlags::values)
+          {
+            std::visit(
+              [&](const auto &obj) {
+                copy_data(
+                  dst.values,
+                  VectorTools::point_values<n_components>(
+                    *obj.rpe, mesh, src, vec_flags, first_selected_component),
+                  obj.get_pntrs());
+              },
+              communication_object);
+          }
+
+        if (eval_flags & EvaluationFlags::gradients)
+          {
+            std::visit(
+              [&](const auto &obj) {
+                copy_data(
+                  dst.gradients,
+                  VectorTools::point_gradients<n_components>(
+                    *obj.rpe, mesh, src, vec_flags, first_selected_component),
+                  obj.get_pntrs());
+              },
+              communication_object);
+          }
+
+        Assert(!(eval_flags & EvaluationFlags::hessians), ExcNotImplemented());
+      }
+
+    if (has_ghost_elements == false)
+      src.zero_out_ghost_values();
+  }
+
+  /**
+   * Provide access to @c PrecomputedFEEvaluationDataView.
+   */
+  const internal::PrecomputedFEEvaluationDataView &
+  get_view() const
+  {
+    return view;
+  }
+
+private:
+  /**
+   * CRS like data structure that describes the data positions at given
+   * indices.
+   */
+  internal::PrecomputedFEEvaluationDataView view;
+
+  /**
+   * A variant for all possible communication objects.
+   */
+  std::vector<std::variant<FERemoteCommunicationObjectEntityBatches<dim>,
+                           FERemoteCommunicationObjectCells<dim>,
+                           FERemoteCommunicationObjectFaces<dim>>>
+    communication_objects;
+
+  // The rest of the class only deals with copying data
```

Comment:
We could create a subsection. Examples are in `Triangulation` or `DoFHandler`.

Reply:
Like this: [8d98c72](https://github.com/dealii/dealii/pull/16300/commits/8d98c726bd47716123d81c528558f05769c72373)?

The diff looks very bad. I simply wrapped the functions into a class and made them static.